src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+{
+	PGresult *result;
+	bool ret = true;
+	va_list args;
+	char buf[MAX_STMT_LEN];
+
+	va_start(args, sql);
+	vsnprintf(buf, sizeof(buf), sql, args);
+	result = PQexec(conn, buf);
+
+	if (PQresultStatus(result) != expectedResult)
+	{
+		ReportRemoteError(conn, result);
+		ret = false;
+	}
+	return ret;
+}
+
+/*
+ * Handle copy to/from distributed table
+ */
+bool PgShardCopy(CopyStmt *copyStatement, char const* query)
+{		 
+	RangeVar *relation = copyStatement->relation;
+	ListCell *shardIntervalCell;
+	ListCell *taskPlacementCell;
+	List *shardIntervalList;
+	int relationSuffixPos;
+	char *relationName;
+	char const *lowerQuery, *relationOcc;
+	PgCopyTransactionManager* tmgr = &PgCopyTmgrImpl[PgCopyTMGR];
+
+	if (relation != NULL)
+	{
+		bool failOK = true;
+		Oid tableId = RangeVarGetRelid(relation, NoLock, failOK);
+		bool isDistributedTable = false;
+
+		isDistributedTable = IsDistributedTable(tableId);
+		if (isDistributedTable)
+		{
+			HTAB* shard2conn;
+			MemoryContext tupleContext;
+			CopyState copyState;
+			bool nextRowFound = true;
+			TupleDesc tupleDescriptor = NULL;
+			uint32 columnCount = 0;
+			Datum *columnValues = NULL;
+			bool *columnNulls = NULL;
+			Var* partitionColumn;
+			Oid columnOid;
+			OpExpr *equalityExpr;
+			HASH_SEQ_STATUS hashCursor;
+			List *whereClauseList;
+			List *prunedList;
+			Node *rightOp;
+			Const *rightConst;
+			int64 failedShard = -1;
+			Relation rel;
+			StringInfo lineBuf;
+			CopyConnection* copyConn;
+			int i;
+
+			relationName = get_rel_name(tableId);
+			lowerQuery = lowerstr(query);
+			relationOcc =  strstr(lowerQuery, lowerstr(relationName));
+			Assert(relationOcc != NULL);
+			relationSuffixPos = (int)((relationOcc - lowerQuery) + strlen(relationName));			 
+
+			shardIntervalList = LookupShardIntervalList(tableId);
+			if (shardIntervalList == NIL)
+			{
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("could not find any shards for query"),
+								errdetail("No shards exist for distributed table \"%s\".",
+										  relationName),
+								errhint("Run master_create_worker_shards to create shards "
+										"and try again.")));
+			}
+			if (!copyStatement->is_from) { /* Construct query selecting data from this relation */
+				uint64 processedCount;
+				char const* qualifiedName = quote_qualified_identifier(relation->schemaname,
+																	   relation->relname);
+				StringInfo newQuerySubstring = makeStringInfo();
+				List *queryList = NIL;
+				appendStringInfo(newQuerySubstring, "select * from %s", qualifiedName);
+				queryList = raw_parser(newQuerySubstring->data);
+				copyStatement->query = linitial(queryList);
+				copyStatement->relation = NULL;
+				/* Collecting data from shards will be done by select handler */
+				DoCopy(copyStatement, query, &processedCount);
+				return true;
+			}
+
+			/* construct pseudo predicate specifying condition for partition key */
+			partitionColumn = PartitionColumn(tableId);
```

Comment:
This seems like boiler-plate for calling PruneShardList. How about adding a ShardsForPartitionValue(partitionValue, shardIntervalSet) function that does the "magic"? This can then be changed to use the fast join pruning logic.